plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts

```diff
@@ -49,23 +47,92 @@ import {
 import { Logger } from 'winston';
 import { PolicyEvaluationRecord } from 'azure-devops-node-api/interfaces/PolicyInterfaces';
-import { WebApi } from 'azure-devops-node-api';
+import {
+  WebApi,
+  getHandlerFromToken,
+  getPersonalAccessTokenHandler,
+} from 'azure-devops-node-api';
 import {
   TeamProjectReference,
   WebApiTeam,
 } from 'azure-devops-node-api/interfaces/CoreInterfaces';
 import { UrlReader } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import {
+  AzureDevOpsCredentialsProvider,
+  DefaultAzureDevOpsCredentialsProvider,
+  ScmIntegrations,
+} from '@backstage/integration';
+import {
+  mappedBuildRun,
+  mappedGitTag,
+  mappedPullRequest,
+  mappedRepoBuild,
+} from './mappers';
 
 /** @public */
 export class AzureDevOpsApi {
-  public constructor(
-    private readonly logger: Logger,
-    private readonly webApi: WebApi,
-    private readonly urlReader: UrlReader,
-  ) {}
+  private readonly logger: Logger;
+  private readonly urlReader: UrlReader;
+  private readonly config: Config;
+  private readonly credentialsProvider: AzureDevOpsCredentialsProvider;
+
+  private constructor(
+    logger: Logger,
+    urlReader: UrlReader,
+    config: Config,
+    credentialsProvider: AzureDevOpsCredentialsProvider,
+  ) {
+    this.logger = logger;
+    this.urlReader = urlReader;
+    this.config = config;
+    this.credentialsProvider = credentialsProvider;
+  }
+
+  static fromConfig(
+    config: Config,
+    options: { logger: Logger; urlReader: UrlReader },
+  ) {
+    const scmIntegrations = ScmIntegrations.fromConfig(config);
+    const credentialsProvider =
+      DefaultAzureDevOpsCredentialsProvider.fromIntegrations(scmIntegrations);
+    return new AzureDevOpsApi(
+      options.logger,
+      options.urlReader,
+      config,
+      credentialsProvider,
+    );
+  }
+
+  private async getWebApi(host?: string, org?: string): Promise<WebApi> {
+    // If no host or org is provided we fall back to the values from the `azureDevOps` config section
+    // these may have been setup in the `integrations.azure` config section
+    // which is why use them here and not just falling back on them entirely
+    const validHost = host ?? this.config.getString('azureDevOps.host');
+    const validOrg = org ?? this.config.getString('azureDevOps.organization');
+    const url = `https://${validHost}/${encodeURI(validOrg)}`;
+
+    const credentials = await this.credentialsProvider.getCredentials({
+      url,
+    });
+
+    let authHandler;
+    if (!credentials) {
+      // No credentials found for the provided host and org in the `integrations.azure` config section
+      // use the fall back personal access token from `azureDevOps.token`
+      const token = this.config.getString('azureDevOps.token');
+      authHandler = getPersonalAccessTokenHandler(token);
+    } else {
+      authHandler = getHandlerFromToken(credentials.token);
+    }
+
+    const webApi = new WebApi(url, authHandler);
```

Comment:
Would it make sense to reconstruct this class and the handlers at most once every N seconds or something like that? Or is there no significant risk / performance impact of reconstructing it over and over for every request?